### ISSUE TITLE
Sparc build reveals interesting artifacts in our code, eg, always true expressions

### ISSUE THREAD

**Issue by demerphq** [Perl/perl5#20479]
**Description**
A bug report was made that referenced this: https://buildd.debian.org/status/fetch.php?pkg=perl&arch=sparc64&ver=5.36.0-4&stamp=1666130893&raw=0 (see #20477)

That ticket is about certain tests failing on Sparc. But the compiler they use appears to do more checks than gcc might and notices a bunch of minor artifacts, such as expressions that are always true. It seems to really not like our habit of sticking a char[1] at the end of a struct (it would be nice to know how to annotate this so the compiler understood the intent), but sprinkled amongst the bogus warnings related to that there are other interesting cases:

```
regcomp.c: In function ‘Perl_re_op_compile’:
regcomp.c:7966:9: warning: the comparison will always evaluate as ‘true’ for the pointer operand in ‘*__builtin_memcpy(&<Uf320>, &({...})) + (sizetype)Perl_ReANY((const REGEXP *)old_re)->pre_prefix’ must not be NULL [-Waddress]
 7966 |         && RX_PRECOMP(old_re)
      |         ^~
```
This is interesting for instance, and it seems to be correct. 

**Steps to Reproduce**
I don't know how to reproduce except to build on sparc.

**Expected behavior**
Perl really should build clean.

**Perl configuration**
I dont have a perl to -V but these are the config_args from that debian report.
```
-Dmksymlinks -Dusethreads -Duselargefiles -Dcc=sparc64-linux-gnu-gcc -Dcpp=sparc64-linux-gnu-cpp -Dld=sparc64-linux-gnu-gcc -Dccflags=-DDEBIAN -DAPPLLIB_EXP="/usr/lib/sparc64-linux-gnu/perl-base" -Wdate-time -D_FORTIFY_SOURCE=2 -g -O2 -ffile-prefix-map=/dummy/build/dir=. -fstack-protector-strong -Wformat -Werror=format-security -Dldflags= -Wl,-z,relro -Dlddlflags=-shared -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=sparc64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.34 -Darchlib=/usr/lib/sparc64-linux-gnu/perl/5.34 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/sparc64-linux-gnu/perl5/5.34 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.34.0 -Dsitearch=/usr/local/lib/sparc64-linux-gnu/perl/5.34.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -Ui_xlocale -Uversiononly -DDEBUGGING=-g -Doptimize=-O2 -dEs -Uuseshrplib
```

**Comment by jkeenan** [comment]
> **Description** A bug report was made that referenced this: https://buildd.debian.org/status/fetch.php?pkg=perl&arch=sparc64&ver=5.36.0-4&stamp=1666130893&raw=0 (see #20477)
> 
> That ticket is about certain tests failing on Sparc. But the compiler they use appears to do more checks than gcc might and notices a bunch of minor artifacts, such as expressions that are always true. It seems to really not like our habit of sticking a char[1] at the end of a struct (it would be nice to know how to annotate this so the compiler understood the intent), but sprinkled amongst the bogus warnings related to that there are other interesting cases:
> 

Is this specific to sparc64?  Is there something here different from what we've already observed with (later) versions of gcc-12 in https://github.com/Perl/perl5/issues/20410?

**Comment by demerphq** [comment]
I doubt its specific to sparc, its just i havent seen it yet on gcc, nor had i noticed your other ticket. Likely they are the same. If you think so then we can go ahead and close this as a dupe.

**Comment by jkeenan** [comment]
The build log provided by @glaubitz appears to be 2 complete build runs, so I manually trimmed the file and ran it through the same program I have used for years and used in https://github.com/Perl/perl5/issues/20410.  I got the same results for this trimmed file as I reported for `gcc-12` on Linux in https://github.com/Perl/perl5/issues/20410#issue-1413753308.
```
$ report-build-warnings gh-20477-sparc64-build-make-test-prep.log.gz 
File:  gh-20477-sparc64-build-make-test-prep.log.gz

  Waddress                                  42
  Warray-bounds                              3
```
So I think that if we can eliminate (or even reduce) the number of build-time warnings reported in that ticket, we'll get the same results on sparc64.  Closing this ticket.
